exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/extended/ExtendedTypeFieldFactory.java

```diff
@@ -0,0 +1,207 @@
+    // Extended types are objects: { "$type": ... }
+    // Extended arrays are [ { "$type": ...
+    TokenIterator tokenizer = fieldDefn.tokenizer();
+    JsonToken token = tokenizer.requireNext();
+    ElementParser parser;
+    switch (token) {
+      case START_OBJECT:
+        parser = extendedTypeParserFor(fieldDefn, false);
+        break;
+      case START_ARRAY:
+        parser = arrayParserFor(fieldDefn);
+        break;
+      default:
+        parser = null;
+    }
+    tokenizer.unget(token);
+    return parser;
+  }
+
+  private ElementParser arrayParserFor(FieldDefn fieldDefn) {
+    TokenIterator tokenizer = fieldDefn.tokenizer();
+    JsonToken token = tokenizer.requireNext();
+    if (token != JsonToken.START_OBJECT) {
+      tokenizer.unget(token);
+      return null;
+    }
+
+    ValueParser element = extendedTypeParserFor(fieldDefn, true);
+    tokenizer.unget(token);
+    if (element == null) {
+      return null;
+    }
+
+    return scalarArrayParserFor(element);
+  }
+
+  private BaseExtendedValueParser extendedTypeParserFor(FieldDefn fieldDefn, boolean isArray) {
+    TokenIterator tokenizer = fieldDefn.tokenizer();
+    JsonToken token = tokenizer.peek();
+    if (token != JsonToken.FIELD_NAME) {
+      return null;
+    }
+
+    String key = tokenizer.textValue().trim();
+    if (!key.startsWith(ExtendedTypeNames.TYPE_PREFIX)) {
+      return null;
+    }
+    return parserFor(fieldDefn, key, isArray);
+  }
+
+  private BaseExtendedValueParser parserFor(FieldDefn fieldDefn, String key, boolean isArray) {
+    switch (key) {
+    case ExtendedTypeNames.LONG:
+      return numberLongParser(fieldDefn, isArray);
+    case ExtendedTypeNames.DECIMAL:
+      return numberDecimalParser(fieldDefn, isArray);
+    case ExtendedTypeNames.DOUBLE:
+      return numberDoubleParser(fieldDefn, isArray);
+    case ExtendedTypeNames.INT:
+      return numberIntParser(fieldDefn, isArray);
+    case ExtendedTypeNames.DATE:
+      return dateParser(fieldDefn, isArray);
+    case ExtendedTypeNames.BINARY:
+    case ExtendedTypeNames.BINARY_TYPE:
+      return binaryParser(fieldDefn, isArray);
+    case ExtendedTypeNames.OBJECT_ID:
+      return oidParser(fieldDefn, isArray);
+    case ExtendedTypeNames.DATE_DAY:
+      return dateDayParser(fieldDefn, isArray);
+    case ExtendedTypeNames.TIME:
+      return timeParser(fieldDefn, isArray);
+    case ExtendedTypeNames.INTERVAL:
+      return intervalParser(fieldDefn, isArray);
+    default:
+      return null;
+    }
+  }
+
+  private BaseExtendedValueParser numberLongParser(FieldDefn fieldDefn, boolean isArray) {
+    return new SimpleExtendedValueParser(
+        fieldDefn.parser(), ExtendedTypeNames.LONG,
+        new StrictBigIntValueListener(loader(),
+            fieldDefn.scalarWriterFor(MinorType.BIGINT, isArray)));
+  }
+
+  private BaseExtendedValueParser numberDecimalParser(FieldDefn fieldDefn, boolean isArray) {
+    // No information about precision and scale, so guess (38, 10).
```

Comment:
Is it possible to obtain information about scale and precision from `fieldDefn`? For the case of the schema provisioning users should be able to specify target scale and precision.